pkg/cloud/services/compute/ports.go

```diff
@@ -0,0 +1,226 @@
+ * add it. We should calculate tags for volumes in the same way we do for other
+ * resources: cluster tags + machine tags. We should also add an additional tag
+ * containing the uid of the kubernetes openstackmachine object. We can add all
+ * of these as server metadata. This will allow us to distinguish in the case
+ * where we find 2 volumes with the same name.
+ */
+
+/*
+	// Ensure we delete the ports we created if we haven't created the server.
+	defer func() {
+		if server != nil {
+			return
+		}
+
+		if err := s.deletePorts(eventObject, portList); err != nil {
+			s.scope.Logger().V(4).Error(err, "Failed to clean up ports after failure")
+		}
+	}()
+
+	ports, err := s.constructPorts(openStackCluster, instanceSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	networkingService, err := s.getNetworkingService()
+	if err != nil {
+		return nil, err
+	}
+
+	securityGroups, err := networkingService.GetSecurityGroups(instanceSpec.SecurityGroups)
+	if err != nil {
+		return nil, fmt.Errorf("error getting security groups: %v", err)
+	}
+
+	for i := range ports {
+		portOpts := &ports[i]
+		iTags := []string{}
+		if len(instanceSpec.Tags) > 0 {
+			iTags = instanceSpec.Tags
+		}
+		portName := networking.GetPortName(instanceSpec.Name, portOpts, i)
+		port, err := networkingService.GetOrCreatePort(eventObject, clusterName, portName, portOpts, securityGroups, iTags)
+		if err != nil {
+			return nil, err
+		}
+
+		portList = append(portList, servers.Network{
+			Port: port.ID,
+		})
+	}
+*/
+
+import (
+	"fmt"
+
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/cluster-api/util/conditions"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha7"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/services/networking"
+	capoerrors "sigs.k8s.io/cluster-api-provider-openstack/pkg/utils/errors"
+)
+
+func (s *Service) adoptPorts(eventObject runtime.Object, clusterName string, openStackMachine *infrav1.OpenStackMachine, instanceSpec *InstanceSpec, desiredPorts []infrav1.PortOpts) error {
+	portsReady := conditions.Get(openStackMachine, infrav1.PortsReadyCondition)
+	if portsReady != nil && portsReady.Status == corev1.ConditionTrue {
+		return nil
+	}
+
+	statusPorts := openStackMachine.Status.Ports
+
+	networkingClient, err := s.getNetworkingClient()
+	if err != nil {
+		return err
+	}
+
+	for i, port := range desiredPorts {
+		// check if the port is in status first and if it is, skip it
+		if i < len(statusPorts) {
+			continue
+		}
+
+		portOpts := &desiredPorts[i]
+		portName := networking.GetPortName(openStackMachine.Name, portOpts, i)
+		// List ports by name and tags
+		ports, err := networkingClient.ListPort(ports.ListOpts{
+			Name:      portName,
+			NetworkID: port.Network.ID,
+		})
+		if err != nil {
+			return err
+		}
+		if len(ports) == 0 {
+			return nil
```

Comment:
@mdbooth I don't get why we returned `nil` here, it'll stop the whole function if a port doesn't exist. So what? I'm trying to figure out how you wanted to find out what ports actually need creation/deletion after adoption.

Reply:
When we write this again we should include some comments!

It will have been because we always create ports in order. This means that if port `N` doesn't exist we know that ports `>N` don't exist.